aggmetric: support high cardinality metrics

aa-joshi opened the pull request:
Previously, we had limited support for high cardinality in CRDB metrics. We use aggmetric to aggregate all combinations of labelValues and export both aggregated and child metrics as per config. This was inadequate because we explicitly have to create the child metric first so that it will be reflected in aggmetric. To address this, this patch add new methods on aggregated metric to support high cardinality. These methods will create a new child metric if we see a new combination of label values. We have replaced children in childset with cache.UnorderedCache. We are using LRU cache with default size of 5000. This would help us in controlling the cardinality for a given metric.

Epic: CRDB-43153
Part of: CRDB-48489
Release note: None

aa-joshi committed:
aggmetric: support high cardinality metrics

Previously, we had limited support for high cardinality in CRDB metrics. We use
aggmetric to aggregate all combinations of labelValues and export both
aggregated and child metrics as per config. This was inadequate because we
explicitly have to create the child metric first so that it will be reflected
in aggmetric. To address this, this patch add new methods on aggregated metric
to support high cardinality. These methods will create a new child metric if we
see a new combination of label values. We have replaced children in childset
with cache.UnorderedCache. We are using LRU cache with default size of 5000.
This would help us in controlling the cardinality for a given metric.

Epic: CRDB-43153
Part of: CRDB-48489
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143323)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143323)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM7fiX62tcIObdNapmn:-OM7fiX62tcIObdNapmo:b-nihfdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/agg_metric.go#L21)):*
> ```Go
> )
> 
> const cacheSize = 5000
> ```

nit: docstring. what is this the size of?
___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM8o3nw51Sgl9vy0qla:-OM7for3-IX0bOmOph2B:bckr979) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/agg_metric.go#L74)):*
> ```Go
> func (cs *childSet) init(labels []string) {
> 	cs.labels = labels
> 	cs.mu.children = cache.NewUnorderedCache(cache.Config{
> ```

You are modifying the implementation of a data structure used in production that will impact its use. Other uses of `AggMetric` assume certain things about its function such as: no children are ever deleted. Can we transition more safely by offering a choice between bounded and unbounded collection?

One thing we can consider is deprecating the unbounded collection in 25.3 once we've introduced the new features and people know how to use them effectively.
___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM7f_6G2A1iyjEKVIIE:-OM7f_6G2A1iyjEKVIIF:berhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/agg_metric.go#L76)):*
> ```Go
> 	cs.mu.children = cache.NewUnorderedCache(cache.Config{
> 		Policy: cache.CacheLRU,
> 		//TODO (aa-joshi) : make it configurable
> ```

nit: avoid abstract "it". Be specific about what needs to be made configurable. 
___
*[`pkg/util/metric/aggmetric/counter.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM8pKDS0wCE9676jc2B:-OM8pKDT79k9rmGtTUz2:bepr8wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/counter.go#L98)):*
> ```Go
> 
> 	// If the child already exists, increment it.
> 	if child, ok := c.get(labelVals...); ok {
> ```

This will allocate a new string containing all the values on every `Inc` invocation. We'll need to find a more efficient way to do it. Alternatively, we could force the caller to maintain a reference to the child metric objects, but I assume we want to avoid that.

I suggest setting up some benchmarks so you can compare the cost of using a labeled metric vs a standard one. Today the cost of both is identical because we force the caller to keep a reference to the child. Ideally the labeled metric doesn't allocate new memory during recording, and incurs a small CPU penalty for computing a key.


<!-- Sent from Reviewable.io -->

aa-joshi committed:
aggmetric: make children metric storage configurable

This patch makes children metric storage in childset configurable. We will use
existing BTree as default. Developer can choose between Btree and
UnorderedCache for storage. Btree storage provides methods to update on child
metrics which will be aggregated to the parent metric. UnorderedCache storage
provides additional methods to perform operations on AggMetric. It will create
the child internally, if required.

Epic: CRDB-43153
Part of: CRDB-48489
Release note: None

aa-joshi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143323)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, @arjunmahishi, and @dhartunian)

___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM8o3nw51Sgl9vy0qla:-OMDS4_j9NBTJtOc-v7d:b8gg7t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/agg_metric.go#L74)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

You are modifying the implementation of a data structure used in production that will impact its use. Other uses of `AggMetric` assume certain things about its function such as: no children are ever deleted. Can we transition more safely by offering a choice between bounded and unbounded collection?

One thing we can consider is deprecating the unbounded collection in 25.3 once we've introduced the new features and people know how to use them effectively.
</blockquote></details>

I have updated PR to support both BTree and Cache storage type. By default, it will be Btree for backward compatibility,
___
*[`pkg/util/metric/aggmetric/counter.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OM8pKDS0wCE9676jc2B:-OMDSJLG-te-35pGHQA5:b-633isv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92757714989980bd9a187da10415e91f186cc5f/pkg/util/metric/aggmetric/counter.go#L98)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This will allocate a new string containing all the values on every `Inc` invocation. We'll need to find a more efficient way to do it. Alternatively, we could force the caller to maintain a reference to the child metric objects, but I assume we want to avoid that.

I suggest setting up some benchmarks so you can compare the cost of using a labeled metric vs a standard one. Today the cost of both is identical because we force the caller to keep a reference to the child. Ideally the labeled metric doesn't allocate new memory during recording, and incurs a small CPU penalty for computing a key.
</blockquote></details>

Yes, We will be doing benchmarking when we update metrics with new definition. The plan is to generate hash based on label values for efficient retrieval of child metric.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-:-OMFCgkG8kghFpBaSp1g:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143323#-OMFCn7B0QCz551IP9kZ:-OMFCn7B0QCz551IP9k_:bhwsi9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf934148859f9b2091601970cc934a08efb6a88/pkg/util/metric/aggmetric/agg_metric.go#L82)):*
> ```Go
> 	// If the cache exceeds this size, the oldest children are evicted. This is
> 	// specific to cache storage for children
> 	const cacheSize = 5000
> ```

sidenote: I'll mention this in the doc for discussion but I'm realizing that when we evict labels in a counter, we might want to aggregate the values into an `_evicted` label so that `sum` over the label still makes sense.


<!-- Sent from Reviewable.io -->

aa-joshi commented:
TFTR!
bors r+
